Enable -Xsource:3 compiler flag

joan38 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Enable -Xsource:3-cross compiler flag for future migration to Scala 3.
https://docs.scala-lang.org/scala3/guides/migration/tooling-scala2-xsource3.html

This PR makes sure this code base compiled with Scala 2.13 aligns with Scala 3. This allows for an easier migration to Scala 3.

It introduces very minor binary incompatibility for APIs that were supposed to be private anyway. The impact of those incompatibilities should be zero. Unless people used those private constructors illegally.

If we exclude the future Scala 3 benefit and strictly focus on the immediate benefits of this PR: it removes wrongfully provided access to private constructors via the `apply` functions in their companion object.

## Why Scala 3?

* Scala 3 was [released in May 2021](https://github.com/scala/scala3/releases/tag/3.0.0) and pretty much the entire community supports it now. This puts the Spark project in an embarrassing position, as it regularly shows up in jokes pointing how Spark is late on upgrading Scala. We can see, we already got reactions on this PR from people hoping to have Scala 3 support one day in Spark.
* Spark showing up in a lot of Scala codebases is causing a lot of pain for users who are also trying to use libraries that can only work on Scala 3.
* Scala 2 has become a complex language, but Scala 3 is more straightforward. It has been redesigned to be significantly simpler, with the removal of complex features considered warts. Some of which have been back ported to Scala 2.13 as we can see in this PR.
* Most learning materials now focus on Scala 3, as Scala 2 is largely being phased out. This leaves Spark users in a bad position to start with Spark, as they need to rely on old learning materials outside of the Spark documentation.
* [Support for Scala 2 is in maintenance mode](https://www.scala-lang.org/blog/2024/12/16/scala-2-maintenance.html). It does not feel right for an advanced piece of tech such as Spark to rely on a language version that is now being phased out.
* Scala 3 ends the binary incompatibility madness. Scala 2.x versions were breaking on every minor versions. As a result, we relied on cross compiling the codebase between those versions. This is painful. Moving to Scala 3 means 1 last break/cross compile and no more breakage ever.
* A potential benefit to Spark could be the more Pythony syntax of Scala 3. Since Spark has a large Python user base, [the optional indentation based syntax](https://docs.scala-lang.org/scala3/reference/other-new-features/indentation.html) might help Python users feel like home with Scala.
* Scala 3 adds a lot of features that catch up with other languages such as [extension methods](https://docs.scala-lang.org/scala3/reference/contextual/extension-methods.html) or [enums](https://docs.scala-lang.org/scala3/reference/enums/enums.html) and more...
* Scala 2 is slow to compile. In our repo Scala 3 is 3x faster to compile than Scala 2 (1min instead of 3min), for the same code!


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
This not only eases potential future Scala 3 migration but also make the compiler stricter with features that have proven to be warts.


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as new features, bug fixes, or other behavior changes. Documentation-only updates are not considered user-facing changes.

If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No


### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Using the existing test suite

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No

Thanks

LuciferYang commented on sql/api/src/main/scala/org/apache/spark/sql/MergeIntoWriter.scala:
```diff
@@ -202,7 +202,7 @@ abstract class MergeIntoWriter[T] {
  *   The type of data in the MergeIntoWriter.
  */
-case class WhenMatched[T] private[sql] (
+case class WhenMatched[T](
```
Such a change has led to an enlargement of the access scope of the constructor, which may not conform to the original intended expectations. Isn't there a better alternative solution?

joan38 replied:
Reverted

LuciferYang commented on project/SparkBuild.scala:
```diff
@@ -229,6 +229,7 @@ object SparkBuild extends PomBuild {
   lazy val compilerWarningSettings: Seq[sbt.Def.Setting[_]] = Seq(
     (Compile / scalacOptions) ++= {
       Seq(
+        "-Xsource:3-cross",
```
Maven also needs to be fixed similarly.

joan38 replied:
Done

joan38 committed:
Enable -Xsource:3-cross compiler flag

This not only eases potential future Scala 3 migration but also make the compiler stricter with features that have proven to be warts.

joan38 commented:
There are [some MiMa failures](https://github.com/joan38/spark/actions/runs/14207689515/job/39809035802#step:13:7578), can someone advice on these?

LuciferYang commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -1014,39 +1014,44 @@ final class ShuffleBlockFetcherIterator(
-                  "prepare to fetch the original blocks", e)
-                pushBasedFetchHelper.initiateFallbackFetchForPushMergedBlock(
-                  shuffleBlockId, pushBasedFetchHelper.localShuffleMergerBlockMgrId)
+        case PushMergedLocalMetaFetchResult(
```
1. Mima check failure indicates that binary forward compatibility has been broken. I think we should not compromise it for the sake of a single compilation parameter. 
2. Additionally, please refrain from modifying the code formatting for sections that have not been actually revised. Our principle is to keep the scope of changes in a patch as minimal as possible. Excluding modules `sql/api`, `sql/connect/common`, `sql/connect/server`, and `sql/connect/client/jvm`, do not use the scalafmt tool to format the code.

joan38 replied:
The indentation is not valide for the Scala 3 compiler even with the brace syntax.

joan38 replied:
```
* static method tupled()scala.Function1 in class org.apache.spark.sql.sources.And does not have a correspondent in current version
  filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.sources.And.tupled")
```
I don't understand how `tupled` got broken by this change. Also I wonder with tasty we should have better binary compat.
I think we need consultancy from compiler people.

LuciferYang replied:
> The indentation is not valide for the Scala 3 compiler even with the brace syntax.

Would this be a compilation warning or a compilation error? Can it be suppressed?

joan38 replied:
@LuciferYang I have some answers:
* I pushed a commit that resolves some incompatibilities unrelated to the `-Xsource:3-cross` compiler flag.
* The remaining incompatibilities are due to the fact that case classes are no more implementing `FunctionX` under this compiler flag that mimics some Scala 3 behavior.

Now the question is: is this a big deal? Did the compiler team decide to break this compatibility because it's a niche thing? We need someone to chime in.

LuciferYang replied:
Since there is no clear plan yet for supporting Scala 3（At least I haven't seen it yet.）, there is indeed a need for more people to participate in the discussion.

joan38 replied:
@LuciferYang before involving more people I played a bit more with the compiler options and came to some more conclusions:
* With `-Xsource:3` and `-Xsource-features:v2.13.15,-case-companion-function` I am able to exclude the "feature" with case classes' companion objects not implementing `FunctionN`.
  That way, we are keeping binary compatibility and we can break that next time we eventually cross compile Scala 3.
* Now there is actually another change which applies the constructor modifiers of case classes to the `apply` (`tupled` and `curried` too).
  This means a case class like `org.apache.spark.paths.SparkPath`:
  ```
  case class SparkPath private (private val underlying: String)
  ```
  will not have any `def apply` in the companion object, whereas so far we did have one, that was accessible publicly. Was this intentional? I don't think so.
  I pushed a commit where I went ahead and added the MiMa exceptions (please review them). However if you tell me "No we cannot break binary compatibility here and we need to let people create SparkPath via the apply function", then another solution is to make the constructor public. A 3rd solution is to keep the constructor private and write our own `public def apply` to make sure we don't break binary compatibility.

Let me know

joan38 replied:
@LuciferYang Let me know what's the preferred way here:
* Keeping the binary compatibility but leaving the public `apply` methods on companion objects with `private` or `protected` constructor accessible.
* Breaking binary compatibility only for people calling the apply methods that should have been private.

LuciferYang replied:
@joan38 Thank you for your work. I will revisit this pr after the official release of Apache  Spark 4.0. 

However, given that there is currently no clear plan to migrate to Scala 3, could you summarize the benefits that this pr can bring? I would like to know some more practical benefits.

joan38 replied:
## Benefits of this PR

This PR makes sure this code base compiled with Scala 2.13 aligns with Scala 3. This allows for an easier migration to Scala 3.
It introduces very minor binary incompatibility for APIs that were supposed to be private anyway. The impact of those incompatibilities should be zero. Unless people used those private constructors illegally.

## Why Scala 3?

* Scala 3 was [released in May 2021](https://github.com/scala/scala3/releases/tag/3.0.0) and pretty much the entire community supports it. This puts the Spark project in an embarrassing position, as it regularly shows up in jokes pointing how Spark is late on upgrading Scala. We can see, we already got reactions on this PR from people hoping to have Scala 3 support one day in Spark.
* Scala 2 has become a complex language, but Scala 3 is more straightforward. It has been redesigned to be significantly simpler, with the removal of complex features considered warts. Some of which have been back ported to Scala 2.13 as we can see in this PR.
* Most learning materials now focus on Scala 3, as Scala 2 is largely being phased out. This leave Spark users in a bad position to start with Spark, as they need to rely on old learning materials outside of the Spark documentation.
* [Support for Scala 2 is in maintenance mode](https://www.scala-lang.org/blog/2024/12/16/scala-2-maintenance.html). It does not feel right for an advanced piece of tech such as Spark to rely on a language that is now being phased out.
* Scala 3 ends the binary incompatibility madness. Scala 2.x versions were breaking on every minor versions. As a result, we relied on cross compiling the codebase between those versions. This is painful. Moving to Scala 3 means 1 last break/cross compile and no more breakage ever.
* A potential benefit to Spark could be the more Pythony syntax of Scala 3. Since Spark has a large Python user base, [the optional indentation based syntax](https://docs.scala-lang.org/scala3/reference/other-new-features/indentation.html) might help Python users feel like home with Scala.

## Conclusion

I don't think we should delay or downplay Scala 3 support. This PR is only a preparation for the codebase and the release of Spark 4 is an occasion to add those really minor binary breaks that should impact no one anyway, since these constructors where supposed to be private.

joan38 replied:
@LuciferYang any update?
cc @dongjoon-hyun @HyukjinKwon @cloud-fan @zhengruifeng

LuciferYang replied:
1. The code for Spark 4.0 was frozen a long time ago, and this pull request  does not comply with the backport rules. 
2. I haven't continued reviewing it or attempting to merge it recently to avoid unnecessary code conflicts before the release of version 4.0. 
3. I still believe that before there is a clear Scala 3 upgrade plan, we should not introduce binary incompatibility due to compilation options or syntax changes. Please note that I am not opposed to upgrading to Scala 3, but there should be a well-defined plan.

joan38 replied:
Ok @LuciferYang. So I'll wait for you guys to process Spark 4 and then you'll ping back on this PR?

In the meantime I'd recommend watching [this talk](https://youtu.be/Qeavi9M65Qw) from @bishabosha especially [the DataFrame demo](https://youtu.be/Qeavi9M65Qw?si=tQxtybnOB9P9GMvk&t=1145) and [the example project](https://github.com/bishabosha/scalar-2025/blob/main/dataframe/test/demo.scala).

Thanks for the help!

joan38 committed:
Make MiMa happy without the -Xsource:3-cross option

joan38 committed:
Make MiMa happy with the -Xsource:3 -Xsource-features:v2.13.15,-case-companion-function options

joan38 committed:
Apply the changes to the new code after rebase

zhengruifeng commented:
also cc @rednaxelafx

He-Pin commented:
Time to move on!

Visorgood commented:
So much looking forward to using Scala 3 in Spark 4.x!

joan38 commented:
@LuciferYang let me know if we are ready to get back on this.

joan38 commented:
@LuciferYang any update?
cc @dongjoon-hyun @HyukjinKwon @cloud-fan @zhengruifeng

joan38 commented:
Can someone explain what is blocking this PR now please?

ekrich commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala:
```diff
@@ -128,17 +128,17 @@ object SparkKubernetesClientFactory extends Logging {
     import scala.language.implicitConversions
-    val Driver = Val(DRIVER_CLIENT_REQUEST_TIMEOUT, DRIVER_CLIENT_CONNECTION_TIMEOUT)
-    val Submission = Val(SUBMISSION_CLIENT_REQUEST_TIMEOUT, SUBMISSION_CLIENT_CONNECTION_TIMEOUT)
+    val Driver = ClientTypeVal(DRIVER_CLIENT_REQUEST_TIMEOUT, DRIVER_CLIENT_CONNECTION_TIMEOUT)
```
These enumerations should probably be put in a `scala-2` source along with adding a `scala-3` source with real enums that are compatible with Java. Since they are inline it might mean some code duplication.

Example: https://github.com/ekrich/sconfig/blob/main/sconfig/shared/src/main/scala-3/org/ekrich/config/ConfigValueType.scala

joan38 replied:
This PR does not cross compile with Scala 3. This would come into a separate PR.
This is a small PR to make Scala 2.13 cross compilable. The big PR actually cross compiling would build on this one.

ekrich commented on sql/api/src/main/scala/org/apache/spark/sql/MergeIntoWriter.scala:
```diff
@@ -250,9 +250,7 @@ case class WhenMatched[T] private[sql] (
-case class WhenNotMatched[T] private[sql] (
-    mergeIntoWriter: MergeIntoWriter[T],
-    condition: Option[Column]) {
+case class WhenNotMatched[T](mergeIntoWriter: MergeIntoWriter[T], condition: Option[Column]) {
```
Consider keeping the original wrapped format.

ekrich replied:
I meant the original code was wrapped so you could keep the extra vertical lines of code minus the `private[sql]`.

joan38 commented:
@ekrich this PR is not cross compiling with Scala 3. It's only preparing the 2.13 codebase to later cross compile Scala 3.

Please refer to [my previous comment](https://github.com/apache/spark/pull/50474#discussion_r2024489759) as to what is breaking according to MiMa. But these are private constructors that were not actually private since we could instantiate the class via the `apply` in the companion object. This should be made private or we should make the class public to make is consistent.

ekrich commented:
That makes total sense. Those discussions about what the make public or not are potentially breaking but decisions should be made. I think you should be very close then if you can get agreement.

Noobzik commented:
Any news regarding this ?
Also please post a JIRA link if this is tracked in jira so we can avoid further "Any news" follow up comment

joan38 commented:
> Without clear evidence of performance improvements and potential new features, it's common to place this kind of upgrade in the "nice to have" category.

I am pointing out to http://github.com/joan38/wick to show case the huge features Scala 3 brings for Spark. Types are more relevant than ever in this AI era where agents code for us. We use types for safety but also to provide context.

So upgrading to Scala 3 would mean unparalleled Spark API improvements. Similar to when Apple released Retina Displays, once you upgrade, you wonder how you used older screens.
